File: src/lib/code/mcs-216/section-1/session-5/4/4.c
#include <stdio.h>

/**
 * Swap two integers
 */
void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

/**
 * Lomuto Partition Scheme
 * @param arr   Array to partition
 * @param low   Starting index
 * @param high  Ending index (pivot is at high)
 * @return      Final pivot index
 */
int partition(int arr[], int low, int high) {
    int pivot = arr[high];  // Select last element as pivot
    int i = low - 1;        // Index of smaller element

    printf("  Partitioning [%d..%d] with pivot=%d\n", low, high, pivot);

    for (int j = low; j < high; j++) {
        printf("    Compare arr[%d]=%d with pivot=%d → ", j, arr[j], pivot);
        if (arr[j] <= pivot) {
            i++;
            if (i != j) {
                swap(&arr[i], &arr[j]);
                printf("SWAP %d ↔ %d\n", arr[i], arr[j]);
            } else {
                printf("No swap needed\n");
            }
        } else {
            printf("Skip (greater)\n");
        }
    }

    // Place pivot in correct position
    swap(&arr[i + 1], &arr[high]);
    printf("  Final swap: place pivot at index %d\n", i + 1);
    return i + 1;
}

/**
 * Recursive Quick Sort
 */
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        printf("quickSort(%d, %d)\n", low, high);
        int pi = partition(arr, low, high);

        // Recursively sort elements before and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

/**
 * Print array
 */
void printArray(int arr[], int size) {
    printf("[");
    for (int i = 0; i < size; i++) {
        printf("%d", arr[i]);
        if (i < size - 1) printf(", ");
    }
    printf("]\n");
}

int main() {
    int arr[] = {12, 20, 22, 16, 25, 18, 8, 10, 6, 15};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("=== Quick Sort Implementation in C ===\n");
    printf("Original: "); printArray(arr, n);
    printf("\n");

    quickSort(arr, 0, n - 1);

    printf("\nSorted:   "); printArray(arr, n);
    return 0;
}

File: src/lib/code/mcs-216/section-1/session-5/4/4.rs
/// Quick Sort Implementation in Rust
/// Uses Lomuto partition scheme with last element as pivot

/// Swaps two elements in a mutable slice
fn swap(arr: &mut [i32], i: usize, j: usize) {
    arr.swap(i, j);
}

/// Partition function
fn partition(arr: &mut [i32], low: usize, high: usize) -> usize {
    let pivot = arr[high];
    let mut i = if low == 0 { 0 } else { low } - 1;

    println!("  Partitioning [{low}..{high}] with pivot={pivot}");

    for j in low..high {
        print!("    Compare arr[{j}]={} with pivot={pivot} → ", arr[j]);
        if arr[j] <= pivot {
            i += 1;
            if i != j {
                swap(arr, i, j);
                println!("SWAP {} ↔ {}", arr[i], arr[j]);
            } else {
                println!("No swap needed");
            }
        } else {
            println!("Skip (greater)");
        }
    }

    // Place pivot in correct position
    swap(arr, i + 1, high);
    println!("  Final swap: place pivot at index {}", i + 1);
    i + 1
}

/// Recursive quick sort
fn quick_sort(arr: &mut [i32], low: usize, high: usize) {
    if low < high {
        println!("quickSort({}, {})", low, high);
        let pi = partition(arr, low, high);
        quick_sort(arr, low, if pi == 0 { 0 } else { pi - 1 });
        if pi + 1 <= high {
            quick_sort(arr, pi + 1, high);
        }
    }
}

fn main() {
    let mut arr = [12, 20, 22, 16, 25, 18, 8, 10, 6, 15];
    let n = arr.len();

    println!("=== Quick Sort Implementation in Rust ===");
    println!("Original: {:?}\n", arr);

    quick_sort(&mut arr, 0, n - 1);

    println!("\nSorted:   {:?}", arr);
}
